### PR TITLE
Add reset button (Incomplete, work needed)

### PR DESCRIPTION
**Pull request by An0n3m0us**
Ref: #8064 
Adds a reset button for the keys.
keys

Here's the process:
1. Mess up the keys.
2. Press the reset button (menu quits)
3. Open "Change Keys" and they are all already reset.

There are a few problems. It's my first time programming in `C++` so some parts of the code are incorrect.

1. The keys text needs to update when the reset button is pressed instead of having to `quitMenu();`
2. Currently, the keys are reset to an array that I created in the loop (see line `290`), instead of getting the list of keys from the settings script or elsewhere. The array isn't initialized properly either.
3. I copied `acceptInput()` and used it as a template, but some of the code might not be needed.

Tasks:
- [x] Make buttons reset correctly.
- [x] Make GUI automatically update when reset button is pressed
- [x] Create an "Are you sure?" popup or double click to make sure.
- [ ] Fix `Segmentation fault (core dumped)` error.

### CODE REVIEW

File: src/gui/guiKeyChangeMenu.cpp [hunk 1]
```diff
@@ -273,6 +283,38 @@ bool GUIKeyChangeMenu::acceptInput()
 	return true;
 }
 
+bool GUIKeyChangeMenu::resetInput()
+{
+	int i = 0;
+	for (key_setting *k : key_settings) {
+		char keys[33][50] = {"W", "S", "A", "D", "E", "KEY_SPACE", "KEY_LSHIFT", "Q", "I", "B", "N", "Z", "KEY_F7", "KEY_F9", "K", "P", "J", "H", "M", "", "", "", "T", "KEY_F12", "R", "-", "+", "KEY_F10", "/", ".", "KEY_F1", "KEY_F2", "KEY_F3"};
```
Comment (An0n3m0us) [inline src/gui/guiKeyChangeMenu.cpp]: Keys should be loaded from some settings file instead of 'recreated'.

File: src/gui/guiKeyChangeMenu.cpp [hunk 2]
```diff
@@ -273,6 +283,38 @@ bool GUIKeyChangeMenu::acceptInput()
...
+		g_settings->set(k->setting_name, keys[i]);
```
Comment (SmallJoker) [inline src/gui/guiKeyChangeMenu.cpp]: These two lines can be replaced with
```c++
g_settings->set(g_settings->getDefault(k->setting_name));

// Or let the engine fall back to the default value by removing the custom setting:
g_settings->remove(k->setting_name);
```

Comment (An0n3m0us) [reply src/gui/guiKeyChangeMenu.cpp]: Like this?:
```
bool GUIKeyChangeMenu::resetInput()
{
	for (key_setting *k : key_settings) {
		g_settings->remove(k->setting_name);
	}
        ...
```

Comment (An0n3m0us) [reply src/gui/guiKeyChangeMenu.cpp]: It worked so I'll add it.

File: src/gui/guiKeyChangeMenu.cpp [hunk 3]
```diff
@@ -377,6 +416,10 @@ bool GUIKeyChangeMenu::OnEvent(const SEvent& event)
 				case GUI_ID_ABORT_BUTTON: //abort
 					quitMenu();
 					return true;
+				case GUI_ID_RESET_BUTTON: //reset
+					resetInput();
+					quitMenu();
```
Comment (An0n3m0us) [inline src/gui/guiKeyChangeMenu.cpp]: Instead of quitting the menu, it needs to be updated somehow.

**Comment by paramat** [comment]
Why is this needed? Keybindings are changed occasionally one by one as required so i can't see much need.

**Comment by An0n3m0us** [comment]
In the case of someone wanting to reset them all, they don't have to delete `minetest.conf`.

**Comment by paramat** [comment]
Ok, but why would they want to reset them all? This would be very rarely needed as players do not suddenly revert their controls to default one day. Changing controls is done one by one gradually.

**Comment by An0n3m0us** [comment]
True.
What about instead having a reset key for each key (or a button to reset each key)? Like escape or backspace?

**Comment by SmallJoker** [comment]
> What about instead having a reset key for each key

Ehem #8282

**Comment by An0n3m0us** [comment]
Oh, oops I didn't come across that.
Is there a need to reset all the buttons at once? Personally, I think it wouldn't do any harm to have one anyway.

**Comment by DS-Minetest** [comment]
The reset all button could be pressed on accident, which might be very annoying.

**Comment by An0n3m0us** [comment]
@DS-Minetest That's why it might need an "Are you sure" popup like I mentioned above.

**Comment by An0n3m0us** [comment]
I've made the GUI automatically update and I've unintentionally added a double click to reset the keys. The only problem is when the player clicks on the keys after resetting them, the game crashes and gives a `segmentation fault`.

**Comment by An0n3m0us** [comment]
Done! The `segmentation fault` was because of something else. Everything seems to work fine now.

**Comment by paramat** [comment]
No progress for 6 months and seems mostly unnecessary.

**Pull request closed without merging**